### DISCUSSION TITLE
Ability to Use External Identity Provider, or Identity Provider Cross-Instance

### DISCUSSION THREAD

**Discussion by atlowell-smpl** [mastodon/mastodon#27952] [General]
I am unsure if this feature exists, so I'm posting it in "support", but if it doesn't, please consider this a feature request.

I am wondering if there is a way to use a third-party identity provider to manage identity for a mastodon instance?  Mastodon could be a very useful platform for expanding an existing application with privacy-focused social features.  However, if users are already identified and authenticated via another service, it would be nice if mastodon could use that authentication directly rather than requiring the application to create and manage mastodon identities on behalf of the user.

This becomes even more complex when you consider the use case of user wanting to join and participate in multiple mastodon instances using a single user account.  It would greatly simplify development if a user's identity could be obtained from the existing identity provider and used across servers.

This leads me to the second part of this issue.  If this isn't possible, or isn't possible in the second use case, is there at least a way to designate a "primary" mastodon instance that other mastodon instances can obtain authentication from?

Or is the design of Mastodon such that identity is inherently tied to a server and cannot be obtained from elsewhere?

For clarification, user in these use cases need not be able to connect to ANY mastodon instance, just ones that are configured to handle authentication from the specified provider.

**Comment by jtayler** [comment]
Please chime in here: https://www.reddit.com/r/Mastodon/comments/17wwlay/hello_trust_and_safety_enthusiasts/

**Comment by ThisIsMissEm** [comment]
Yes, you can configure your instance to use login via OpenID, PAM and LDAP, it's not properly documented at the moment, but is possible: https://docs.joinmastodon.org/admin/config/#external-authentication

However, I think what you're describing is that the account wouldn't be owned by mastodon, which isn't supported in mastodon's architecture at present.

**Comment by jtayler** [comment]
Nice! Multi-factor login and security is important so we know the person using the account is really the owner. The question we're trying to ask/answer in that discussion is really "are you who you say" kind of identity which is a related facet of identity.

**Reply by ThisIsMissEm** [reply]
The "are you who you say you are" part of identity is best solved by verified URLs: https://joinmastodon.org/verification

**Comment by martinmodrak** [comment]
I am considering options for integrating Mastodon with an identity provider that uses its own protocol (i.e. I cannot use OpenID/PAM/LDAP and would need to write some of my own code for the authentication flow). Do I understand correctly that to do this I would have to fork Mastodon? Or is there some way to sneak in an "identity provider plugin"?

**Comment by jtayler** [comment]
TruAnon has been integrated with Mastodon quite nicely, right now you need to pull that code into your server but it is supposed to be a validation system that supports whatever services you might want to offer.

Im happy to help get that done somehow if I am able!
